src/Common/tests/Collections/ICollectionTest.cs

```diff
@@ -0,0 +1,288 @@
+                    () => collection.CopyTo(itemArray, -1));
+                CollectionAssert.Equal(tempItemArray, itemArray);
+            }
+        }
+
+        [Fact]
+        public void CopyToIndexArrayLength()
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            var tempArray = (object[]) items.Clone();
+            Assert.Throws<ArgumentException>(
+                () => collection.CopyTo(items, collection.Count));
+            CollectionAssert.Equal(tempArray, items);
+        }
+
+        [Fact]
+        public void CopyToCollectionLargerThanArray()
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            var itemArray = GenerateItems(collection.Count + 1);
+            var tempItemArray = (object[]) itemArray.Clone();
+            Assert.Throws<ArgumentException>(
+                () => collection.CopyTo(itemArray, 2));
+            CollectionAssert.Equal(tempItemArray, itemArray);
+        }
+
+        [Fact]
+        public void CopyToMDArray()
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            Assert.Throws<ArgumentException>(
+                () =>
+                collection.CopyTo(new object[1, collection.Count], 0));
+        }
+
+        protected void AssertThrows(
+            Type[] exceptionTypes,
+            Action testCode)
+        {
+            var exception = Record.Exception(testCode);
+            if (exception == null)
+                throw new AssertActualExpectedException(exceptionTypes, null, "Expected an exception but got null.");
+            var exceptionType = exception.GetType();
+            if (!exceptionTypes.Contains(exceptionType))
+                throw new AssertActualExpectedException(exceptionTypes, exceptionType, "Caught wrong exception.");
+        }
+
+        [Fact]
+        public void CopyToInvalidTypes()
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            Type[] expectedExceptionTypes = IsGenericCompatibility
+                                                ? new[]
+                                                {
+                                                    typeof (
+                                                      ArgumentException),
+                                                    typeof (
+                                                      InvalidCastException
+                                                      )
+                                                }
+                                                : new[]
+                                                {
+                                                    typeof (
+                                                      ArgumentException)
+                                                };
```

Comment:
That's an... interesting style :)

It could probably use some autoformatter love.